pkg/kubelet/logs/container_log_manager.go

```diff
@@ -176,19 +176,108 @@ func (c *containerLogManager) Start() {
 	}, logMonitorPeriod)
 }
 
+type containerWithLog struct {
+	ID        string
+	CreatedAt int64
+	logPath   string
+	State     runtimeapi.ContainerState
+}
+
+func (c *containerLogManager) addContainerToMap(container *runtimeapi.Container, path string, containersByName map[string][]containerWithLog) {
+	if container.Metadata != nil {
+		key := container.Metadata.Name + " " + container.PodSandboxId
+		for i := range containersByName[key] {
+			if containersByName[key][i].ID == container.Id {
+				containersByName[key][i].ID = container.Id
+				containersByName[key][i].State = container.State
+				containersByName[key][i].CreatedAt = container.CreatedAt
+				if containersByName[key][i].logPath == "" {
+					containersByName[key][i].logPath = path
+				}
+				return
+			}
+		}
+		withLog := containerWithLog{
+			ID:        container.Id,
+			State:     container.State,
+			CreatedAt: container.CreatedAt,
+			logPath:   path,
+		}
+		containersByName[key] = append(containersByName[key], withLog)
+	}
+}
+
+func (c *containerLogManager) pruneDeadContainerLogs(containersByName map[string][]containerWithLog) {
+	containersToKeep := 2
+	for k, containers := range containersByName {
+		sort.Sort(containersByCreation(containers))
+		var logPath string
+		// we calculate the prefixes of logs to keep for containersToKeep containers
+		prefixesToKeep := make([]string, 0)
+		for i := range containers {
+			if i >= containersToKeep {
+				klog.V(4).Infof("not keeping log for container %s / %s", k, containers[i].ID)
+				continue
+			}
+			if containers[i].logPath != "" {
+				lastSeparator := strings.LastIndex(containers[i].logPath, string(os.PathSeparator))
+				if lastSeparator < 0 {
+					klog.Warningf("cannot find separator in %q", containers[i].logPath)
+					continue
+				}
+				runes := []rune(containers[i].logPath)
+				if logPath == "" {
+					// include the trailing separator
+					logPath = string(runes[0 : lastSeparator+1])
+				} else if containers[i].logPath != "" && !strings.HasPrefix(containers[i].logPath, logPath) {
+					klog.Warningf("%q doesn't begin with %q", containers[i].logPath, logPath)
+				}
+				prefixesToKeep = append(prefixesToKeep, string(runes[lastSeparator+1:]))
+			}
+		}
+		klog.V(4).Infof("cleaning log for %s, path %s", k, logPath)
+		pattern := fmt.Sprintf("%s*", logPath)
+		logs, err := filepath.Glob(pattern)
+		if err != nil {
+			klog.Warningf("failed to list all log files with pattern %q: %v", pattern, err)
+			continue
+		}
+		i := 0
+	outer:
+		for ; i < len(logs); i++ {
+			if !strings.Contains(logs[i], ".log") {
```

Comment:
strings.HasSuffix

Reply:
Thanks for taking a look, @rphillips 

The cleanup needs to handle log files in the following form (one example):
```
/var/log/pods/default_log-test_267032c4-7002-41d8-9420-f7c07a48e898/log-test/1.log.20200424-200024.gz
```
Hence I use Contains() instead of strings.HasSuffix()